### ISSUE TITLE
Figure2

### ISSUE THREAD

**Issue by Curiosity-Mars** [Curiosity-Mars/total-testosterone#6]
# ============================================================
# Figure 2 (Rebuild): Elbow (SSE) + Silhouette vs K
# - Pipeline一致：Subjects= age & BMI complete (N=5854)
# - その後：mean imputation -> standardization -> KMeans
# - 出力：Figure2_elbow_silhouette.png / .pdf + CSV + ZIP download
# ============================================================

import os
import numpy as np
import pandas as pd
import zipfile
import matplotlib.pyplot as plt

from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score

# Colab download
try:
    from google.colab import files
    IN_COLAB = True
except Exception:
    IN_COLAB = False

# -------------------------
# 0) I/O settings
# -------------------------
XLSX_PATH = "/content/drive/MyDrive/CommmedpatientTT.xlsx"  # <- adjust if needed
assert os.path.exists(XLSX_PATH), f"File not found: {XLSX_PATH}"

OUTDIR = "/content/fig2_outputs"
os.makedirs(OUTDIR, exist_ok=True)

# -------------------------
# 1) Load + column harmonization
# -------------------------
df = pd.read_excel(XLSX_PATH)

target_cols = [
    "age","height","BW","BMI","Hb","Hct","Alb","AST","ALT","ALP","Cre",
    "Glu","TG","HDL-C","LDL-C","LH","FSH","CRP","TT"
]

rename_map = {
    "HDL": "HDL-C", "LDL": "LDL-C",
    "HDL_C": "HDL-C", "LDL_C": "LDL-C",
    "BodyWeight": "BW", "Weight": "BW",
}
df = df.rename(columns={k: v for k, v in rename_map.items() if k in df.columns})

missing_cols = [c for c in target_cols if c not in df.columns]
assert len(missing_cols) == 0, f"Missing required columns: {missing_cols}"

X_raw = df[target_cols].copy()

# -------------------------
# 2) [Optional, A要素] BMI reconstruction from height/BW (fill BMI only where missing)
#     ※height欠損なら埋まらないので自然に残る
# -------------------------
h = pd.to_numeric(X_raw["height"], errors="coerce")
w = pd.to_numeric(X_raw["BW"], errors="coerce")
height_is_cm = h.median(skipna=True) > 3.0
h_m = h / 100.0 if height_is_cm else h
bmi_from_hw = w / (h_m ** 2)

mask_bmi_na = X_raw["BMI"].isna()
X_raw.loc[mask_bmi_na, "BMI"] = bmi_from_hw.loc[mask_bmi_na]

# -------------------------
# 3) Subjects filter: complete age & BMI => N=5854 (expected)
# -------------------------
mask_subj = X_raw["age"].notna() & X_raw["BMI"].notna()
X_subj = X_raw.loc[mask_subj].copy()
N_SUBJ = len(X_subj)
print("Subjects (age & BMI complete):", N_SUBJ)
print("height unit assumed:", "cm" if height_is_cm else "m")

# -------------------------
# 4) Mean impute -> standardize
# -------------------------
imputer = SimpleImputer(strategy="mean")
scaler = StandardScaler()

X_imp = imputer.fit_transform(X_subj.values)
X_std = scaler.fit_transform(X_imp)

# -------------------------
# 5) Compute SSE (inertia) + silhouette for K range
#    - silhouetteは計算コストがあるのでKを絞るのが普通
# -------------------------
K_MIN, K_MAX = 2, 10   # <- 必要なら上限を増やしてOK
RANDOM_STATE = 0
N_INIT = 50

rows = []
for k in range(K_MIN, K_MAX + 1):
    km = KMeans(n_clusters=k, n_init=N_INIT, random_state=RANDOM_STATE)
    labels = km.fit_predict(X_std)

    sse = float(km.inertia_)
    sil = float(silhouette_score(X_std, labels)) if k >= 2 else np.nan

    rows.append({"K": k, "SSE": sse, "silhouette": sil})

metrics = pd.DataFrame(rows)
metrics_path = os.path.join(OUTDIR, "fig2_elbow_silhouette_metrics.csv")
metrics.to_csv(metrics_path, index=False)
print("Saved metrics:", metrics_path)
display(metrics)

# -------------------------
# 6) Plot Figure 2 (SSE + silhouette)
#    - 1枚のFigureに上下2段（Elbowとsilhouette）でまとめるのが論文向き
#    - 凡例は不要（ユーザー希望に合わせてlegend無し）
# -------------------------
fig = plt.figure(figsize=(8, 8), dpi=200)

# (A) Elbow: SSE vs K
ax1 = plt.subplot(2, 1, 1)
ax1.plot(metrics["K"], metrics["SSE"], marker="o")
ax1.set_xlabel("Number of clusters (K)")
ax1.set_ylabel("SSE (within-cluster sum of squares)")
ax1.set_title("Figure 2a. Elbow method (SSE)")

# (B) Silhouette vs K
ax2 = plt.subplot(2, 1, 2)
ax2.plot(metrics["K"], metrics["silhouette"], marker="o")
ax2.set_xlabel("Number of clusters (K)")
ax2.set_ylabel("Average silhouette coefficient")
ax2.set_title("Figure 2b. Silhouette analysis")

plt.tight_layout()

png_path = os.path.join(OUTDIR, "Figure2_elbow_silhouette.png")
pdf_path = os.path.join(OUTDIR, "Figure2_elbow_silhouette.pdf")
plt.savefig(png_path, bbox_inches="tight")
plt.savefig(pdf_path, bbox_inches="tight")
plt.show()

print("Saved figure:", png_path)
print("Saved figure:", pdf_path)

# -------------------------
# 7) Save meta + ZIP -> download
# -------------------------
meta_path = os.path.join(OUTDIR, "fig2_meta.csv")
pd.DataFrame({
    "N_subjects": [N_SUBJ],
    "height_unit_assumed": ["cm" if height_is_cm else "m"],
    "K_range": [f"{K_MIN}-{K_MAX}"],
    "kmeans_random_state": [RANDOM_STATE],
    "kmeans_n_init": [N_INIT],
    "imputation": ["mean"],
    "standardization": ["z-score (StandardScaler)"],
}).to_csv(meta_path, index=False)

zip_path = os.path.join(OUTDIR, "fig2_outputs.zip")
with zipfile.ZipFile(zip_path, "w", compression=zipfile.ZIP_DEFLATED) as zf:
    for fname in [
        "Figure2_elbow_silhouette.png",
        "Figure2_elbow_silhouette.pdf",
        "fig2_elbow_silhouette_metrics.csv",
        "fig2_meta.csv",
    ]:
        zf.write(os.path.join(OUTDIR, fname), arcname=fname)

print("ZIP created:", zip_path)

if IN_COLAB:
    files.download(zip_path)
else:
    print("Not in Colab: download manually from:", zip_path)
# ============================================================
